### PR TITLE
[stable/rabbitmq-ha] Add RabbitMQ Prometeus Plugin

### PR DESCRIPTION
**Pull request by Gsantomaggio**
-->
#### Is this a new chart
No
#### What this PR does / why we need it:
With rabbitmq 3.8 the rabbitmq prometheus plugin is shipped by default
With this commit is possible to enable/disable it

#### Special notes for your reviewer:
This plugin is shipped only with RabbitMQ 3.8.
It is useful for people (like me ) that need to use RabbitMQ 3.8.
Thank you

cc @steven-sheehy 
#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

Signed-off-by: Gabriele Santomaggio <g.santomaggio@gmail.com>

### CODE REVIEW

**Comment by steven-sheehy** [comment]
@zanhsieh Please stop ok-to-test PRs you are not reviewing. PRs should be reviewed and feedback applied before testing to not waste limited CI resources.
/assign @steven-sheehy

**Comment by Gsantomaggio** [comment]
cc @gerhard @michaelklishin

**Comment by gerhard** [comment]
Where should we mention the Grafana dashboards that go with this?

https://grafana.com/orgs/rabbitmq (more on the way)

**Comment by Gsantomaggio** [comment]
@gerhard I'd wait for feedback about this PR from the maintainers.
After that maybe we will add the info about Grafana, maybe we can add some note to the value.yaml file

File: stable/rabbitmq-ha/values.yaml
Comment (steven-sheehy) [inline stable/rabbitmq-ha/values.yaml]: Should be `rabbitmqPrometheusPlugin`

File: stable/rabbitmq-ha/Chart.yaml
Comment (steven-sheehy) [inline stable/rabbitmq-ha/Chart.yaml]: `1.34.0` for new features

File: stable/rabbitmq-ha/README.md
Comment (steven-sheehy) [inline stable/rabbitmq-ha/README.md]: Add a note to the prometheus section at the bottom in readme about if you're using RabbitMQ 3.8 you should not use `prometheus.exporter.enabled=true` but use the plugin instead.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]
So you've tested this chart with RabbitMQ 3.8 and it works fine with only this change necessary? If so, that's good to hear. We should probably bump the RabbitMQ version in the near future.

File: stable/rabbitmq-ha/values.yaml
Comment (Gsantomaggio) [reply stable/rabbitmq-ha/values.yaml]: Ok, will change

File: stable/rabbitmq-ha/Chart.yaml
Comment (Gsantomaggio) [reply stable/rabbitmq-ha/Chart.yaml]: Perfect

File: stable/rabbitmq-ha/README.md
Comment (Gsantomaggio) [reply stable/rabbitmq-ha/README.md]: makes sense :)!

**Comment by Gsantomaggio** [comment]
> So you've tested this chart with RabbitMQ 3.8 and it works fine with only this change necessary?
 
I'd say yes, this is deployed in a k8s cluster. I used my branch to test it.
 
Screenshot from 2019-10-14 17-16-40


and here the metrics:

metrics

**Comment by Gsantomaggio** [comment]
@steven-sheehy The PR is ready with the changes! 
Let me know if everything is ok.

test:
```
 helm install . --name rmq-ha --set rabbitmqPrometheusPlugin.enabled=true --set image.tag=3.8.0-alpine --set rabbitmqPassword="test"
```
result (added the metrics port):

Screenshot from 2019-10-15 11-47-20

**Comment by steven-sheehy** [comment]
@Gsantomaggio You need to fix the DCO by signing your commits

In your opinion, should we remove the exporter sidecar option once the chart starts using 3.8? Do you know if the metrics the exporter provides and the plugin provides are the same or vary greatly?

**Comment by Gsantomaggio** [comment]
@steven-sheehy sorry for the missing. 
it should be ok now.
>in your opinion, should we remove the exporter sidecar option once the chart starts using 3.8?

do you mean [repository: kbudde/rabbitmq-exporter](https://github.com/helm/charts/blob/master/stable/rabbitmq-ha/values.yaml#L483) ?

**Comment by gerhard** [comment]
> In your opinion, should we remove the exporter sidecar option once the chart starts using 3.8? Do you know if the metrics the exporter provides and the plugin provides are the same or vary greatly?

I think that we should allow for a transition period of several months, before removing the sidecar in favour of `rabbitmq_prometheus`. https://github.com/rabbitmq/rabbitmq-prometheus/issues/9 has some details on the missing metrics from rabbitmq_prometheus.

Our goal is to close the gap and make both [kbudde/rabbitmq-exporter](https://github.com/kbudde/rabbitmq_exporter) & [deadtrickster/prometheus_rabbitmq_exporter](https://github.com/deadtrickster/prometheus_rabbitmq_exporter) redundant. They both depend on metrics coming from [rabbitmq/rabbitmq-management](https://github.com/rabbitmq/rabbitmq-management) & [rabbitmq/rabbitmq-management-agent](https://github.com/rabbitmq/rabbitmq-management-agent) which we plan to eventually stop supporting. We will be talking about this at [RabbitMQ Summit](https://rabbitmqsummit.com) in November 2019.

Also relevant to this conversation: https://github.com/prometheus/docs/pull/1414

**Comment by steven-sheehy** [comment]
@gerhard Thanks for the response. Sounds like we should support both for the foreseeable future as people have alerts and dashboards built off the `kbudde/rabbitmq-exporter` and the metrics probably vary. Later once the RabbitMQ Prometheus Plugin is more widespread we can consider removing the exporter option.

@Gsantomaggio I think we need to document in the README the 2 different metrics approaches that there now are as part of this PR. Can you create a `Monitoring` section with 2 sub-sections: `Prometheus Exporter` and `RabbitMQ Prometheus Plugin`? Mention how these 2 are mutually exclusive and what setting to enable each. Here you can also add a link to the plugin's dashboard that gerhard mentioned.

**Comment by gerhard** [comment]
> @gerhard Thanks for the response. Sounds like we should support both for the foreseeable future as people have alerts and dashboards built off the `kbudde/rabbitmq-exporter` and the metrics probably vary. Later once the RabbitMQ Prometheus Plugin is more widespread we can consider removing the exporter option.

[We share dashboards publicly](https://grafana.com/orgs/rabbitmq) - [RabbitMQ Overview](https://grafana.com/grafana/dashboards/10991) is the most notable one, and we intend on following-up with alerts shortly. There will be a maturing process, for sure, and we are missing some metrics, but all metrics that are exposed currently have the least impact on the cluster as a whole. Metrics are node local (no server-side aggregation) and they will continue being emitted even when nodes are under extreme pressure. That's possibly the biggest advantage of using the new built-in plugin.  [Monitoring with Prometheus & Grafana](https://www.rabbitmq.com/prometheus.html) guide has a lot more detail to the new approach.

> Mention how these 2 are mutually exclusive and what setting to enable each.

Both can be enabled at the same time, they are not mutually exclusive.

**Comment by Gsantomaggio** [comment]
@steven-sheehy Done.

Thank you to @gerhard for helping in this issue and for provide the  documentation.

**Review by batazor — APPROVED** [review]

File: stable/rabbitmq-ha/README.md
Comment (steven-sheehy) [inline stable/rabbitmq-ha/README.md]: Remove this line. Yes, technically they can be ran at the same time since they're different ports but my point about mutually exclusivity was that you wouldn't generally want to run both at the same time and it could cause incorrect metrics if they both update a metric with the same name.

File: stable/rabbitmq-ha/README.md
Comment (steven-sheehy) [inline stable/rabbitmq-ha/README.md]: `traditional rabbitmq_exporter sidecar container`. Also make rabbitmq_exporter a link to https://github.com/kbudde/rabbitmq_exporter

File: stable/rabbitmq-ha/values.yaml
Comment (steven-sheehy) [inline stable/rabbitmq-ha/values.yaml:261]: I don't see how this config is used anywhere.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: stable/rabbitmq-ha/values.yaml
Comment (Gsantomaggio) [reply stable/rabbitmq-ha/values.yaml:261]: sorry you are right, fixed here https://github.com/helm/charts/pull/17925/commits/4ff4bf79367310d2f7e43ae8a997ee57fda7064b#diff-28e0ed5d78530ec2e1cf87dab1eb62a5R115-R119

File: stable/rabbitmq-ha/README.md
Comment (Gsantomaggio) [reply stable/rabbitmq-ha/README.md]: fixed here: https://github.com/helm/charts/pull/17925/files#diff-a03c73b511fa43f47f906cf3fb1677d5R266

File: stable/rabbitmq-ha/README.md
Comment (Gsantomaggio) [reply stable/rabbitmq-ha/README.md]: I removed the section, but let me say that native and HTTP are two different kinds of monitoring. 
>it could cause incorrect metrics 

I'd say  "incorrect reading"  instead of "incorrect metrics ":)! 
btw I agree with you, you have to know that native is for single node and http is for whole the cluster.

Comment (steven-sheehy) [reply stable/rabbitmq-ha/README.md]: Not sure what you mean by native vs HTTP. Native in the sense that it's built into RabbitMQ as a plugin, I guess. But they're both HTTP endpoints that Prometheus scrapes.

File: stable/rabbitmq-ha/README.md
Comment (steven-sheehy) [inline stable/rabbitmq-ha/README.md]: `traditional`

**Review by steven-sheehy — CHANGES_REQUESTED** [review]
One small spelling error and it looks good

File: stable/rabbitmq-ha/README.md
Comment (gerhard) [reply stable/rabbitmq-ha/README.md]: Yes, native in this context means RabbitMQ plugin, which runs within the same system process as RabbitMQ, the Erlang VM.

Comment (gerhard) [reply stable/rabbitmq-ha/README.md]: rabbitmq_exporter gets the metrics from the RabbitMQ Management HTTP API, and then serves them from its own TCP listener.

rabbitmq-prometheus, the plugin, runs and reads metrics directly from within the Erlang VM where RabbitMQ runs, and exposes them via a dedicated TCP listener. This TCP listener is owned by the same system process that runs RabbitMQ - it's the same Erlang VM. All this is wrapped in the `native` claim.

Would you agree @brian-brazil?

**Review by steven-sheehy — APPROVED** [review]
/lgtm
Thanks!

File: stable/rabbitmq-ha/README.md
Comment (brian-brazil) [reply stable/rabbitmq-ha/README.md]: That sounds about right to me. Using both shouldn't cause a problem, though I'd probably clarify that you usually only want to scrape one of them.

**Pull request merged**
